### ISSUE TITLE
[BUG] Lightning Address not recognized

### ISSUE THREAD

**Issue by ZoltanAB** [lnbits/lnbits#1638]
Lightning Address not recognized.

For payments to lightning addresses, legend.lnbits.com service is not following redirects I believe. For example if the lightning address is a@b.com it looks for https://b.com/.well-known/lnurlp/a. Many websites redirect non-www requests to www, so it gets redirected to https://www.b.com/.well-known/lnurlp/a. This is not followed. So, even if the final version (the one with www), has the correct callback json, legend.lnbits.com does not recognize the address as a valid lightning address.

Most other wallets I tested (Wallet of Satoshi. LifPay, Spark, etc.) follow redirects so I believe legend.lnbits.com should also follow them. IMO, at least 2 or 3 redirects should be followed for best user experience.

**Comment by Perlover** [comment]
And what does the redirect to the subdomain 'www' have to do with it at all? Why should LNBits take into account site redirects? If any site is configured so that the domain b.com redirects to www.b.com - this is a problem of setting up the site, not LNBits or any other client working with Lightning Address.

**Comment by Perlover** [comment]
If we talk in principle about the cases when the owners of the domain b.com want to use the address a@b.com without having software that handles work with Lightning Address (or when they want NIP-57 support), then I have seen another practice:

Lightning Address for Nostr network: vitor@vitorpamplona.com

Map it to: https://vitorpamplona.com/.well-known/lnurlp/vitor

See there:

```
{
	"status": "OK",
	"tag": "payRequest",
	"commentAllowed": 255,
	"callback": "https://getalby.com/lnurlp/vitorpamplona/callback",
	"metadata": "[[\"text/identifier\",\"vitorpamplona@getalby.com\"],[\"text/plain\",\"Sats for Vitor Pamplona\"]]",
	"minSendable": 1000,
	"maxSendable": 11000000000,
	"payerData": {
		"name": {
			"mandatory": false
		},
		"email": {
			"mandatory": false
		}
	},
	"nostrPubkey": "79f00d3f5a19ec806189fcab03c1be4ff81d18ee4f653c88fac41fe03570f432",
	"allowsNostr": true
}
```

That is, if you want to do something like this, then you need to put a callback from another domain.

**Comment by ZoltanAB** [comment]
> And what does the redirect to the subdomain 'www' have to do with it at all? Why should LNBits take into account site redirects? If any site is configured so that the domain b.com redirects to [www.b.com](http://www.b.com) - this is a problem of setting up the site, not LNBits or any other client working with Lightning Address.

I thought that the main goal is to increase adoption. If tomorrow google.com, NYTimes.com, BBC.com (just to name a few big sites), decide to enable Lightning Address support to all their users / employees, do you think that they will change the way their sites are working and no longer redirect anything that doesn't start with www to the www version? No, they won't, we should adapt, if we want lightning address to be a standard.

And at current setup a lightning address of bob@nytimes.com will not be recognized by lnbits just because https://nytimes.com/.well-known/lnurlp/bob is automatically 301 redirected to the www equivalent.

Not to mention the various changes in lightning development when you setup a domain to domain1.io, build a community of tens, hundreds or thousands of users (user@domain1.io) but later decide to move to another domain (domain2.io). What do you do? Inform your users to "go away, domain1.io is no longer supported"? No, most normal businesses keep the domain name for a while and simply redirect all domain1.io to domain2.io and slowly inform their users to get rid of all legacy addresses.

**Comment by ZoltanAB** [comment]
> 

I am aware of this and in fact I am using it, I was not talking about callback. Check above.

**Comment by ZoltanAB** [comment]
If lnbits uses curl to get the json of a lightning address, it is just a simple option to be added to curl to follow a certain number of redirects.

curl -L --max-redirs 3 ...

**Comment by Perlover** [comment]
Trusting redirects is a dangerous thing. What you are suggesting is not even safe. For example, if what you are asking for was implemented, then I could create an address on my domain that would point a redirect to your address. Next, I distribute this address as my own, the user sees my domain, but the payment goes to you. This is a misconception of users. I think scammers will figure out how to exploit it ;)

**Comment by Perlover** [comment]
Your motivation to implement a redirect just because some sites make a redirect to www is some kind of stupid crutch idea. To begin with, you are talking about sites that you apparently do not control. But at the same time, you expect that these sites will allow you to place JSON data in the ".well-known" directory, which will never be available to ordinary users. If the site decides to implement Lightning Address meaningfully for users, they will certainly make sure that there is no redirect for such domains to the www subdomain (otherwise they will be just idiots, and it is definitely dangerous for them to have a Lightning Address in this case). So, why is it necessary for everyone to implement redirects where they are not even needed because of some idiots?

**Comment by ZoltanAB** [comment]
Fine. If you consider anyone who follows 20+ years old SEO strategies idiots, then I have nothing more to say.

Just one last thing. Please try to tip saylor@microstrategy.com from your legend.lnbits wallet. Let me tell you, it won't work. But it will work from WOS, LifPay, Spark, etc.

https://wheregoes.com/trace/20231892685/

**Comment by Perlover** [comment]
In general, in truth, this is a problem of incorrect site settings. By and large, this could have been decided by Sailor himself, since if he had rights to the `.well-known` directory, he also has rights to change the redirect algorithm. This is solved by a couple of lines in the server configuration file.

Well, if he can't do it, he's too lazy, then let him write a new address: `saylor@www.microstrategy.com`

Technically speaking, the 301 redirect says that my address has been PERMANENTLY MOVED to such a new address. That is purely technically it should be interpreted as follows: my address is not saylor@microstrategy.com , but saylor@www.microstrategy.com and go there. And, by the way, this address works successfully in LNBits: saylor@www.microstrategy.com. Who knows, maybe that's what he wanted to say. And if not, well, that's his problem. ;-)

**Comment by Perlover** [comment]
I sent to him to address saylor@www.microstrategy.com the 1234 sats from my LNBits instance with comment of this URL :-)

**Comment by Perlover** [comment]
https://github.com/lnurl/luds/blob/luds/16.md

I looked at the lightning address standard and there is nothing about redirects there. That is, according to the LNBits standard, it works correctly.

**Comment by Perlover** [comment]
Of course, it would be better if LUD-16 described more clearly whether redirects can be used or not. In clearer standards, such as RFC, such details are important. Everything is written here in general, so everyone interprets and implements as he wants. For example, you offered a maximum of 3 redirects, but why so many? Why not 5, or 10? And what if others implement a maximum of 5, others 3, and someone 1? In all cases, there will be situations where everything will work differently. In this context, LNBits has the number 0. What number other wallets that allow redirects have is known only to them.

**Comment by ZoltanAB** [comment]
> In general, in truth, this is a problem of incorrect site settings. By and large, this could have been decided by Sailor himself, since if he had rights to the `.well-known` directory, he also has rights to change the redirect algorithm. This is solved by a couple of lines in the server configuration file.
> 
> Well, if he can't do it, he's too lazy, then let him write a new address: `saylor@www.microstrategy.com`
> 
> Technically speaking, the 301 redirect says that my address has been PERMANENTLY MOVED to such a new address. Moreover, client software usually remembers the destination address of the 301 redirect and no longer addresses the original address. That is, if the Sailor decides to change the callback or something else, then it will only work for those who have never sent it to him yet. And those who sent - they will not be able to receive data for `saylor@microstrategy.com` only for `saylor@www.microstrategy.com`.
> 
> That is purely technically it should be interpreted as follows: my address is not [saylor@microstrategy.com](mailto:saylor@microstrategy.com) , but [saylor@www.microstrategy.com](mailto:saylor@www.microstrategy.com) and go there. And, by the way, this address works successfully in LNBits: [saylor@www.microstrategy.com](mailto:saylor@www.microstrategy.com). Who knows, maybe that's what he wanted to say. And if not, well, that's his problem. ;-)

That's the point. "This could have been decided by Sailor himself, since if he had rights to the .well-known directory,"

If a site owner decides that this is what he wants to do, it is his right to do so. It could have been a a 302 redirect (temporary), but he decided to be a 301.

> I sent to him to address [saylor@www.microstrategy.com](mailto:saylor@www.microstrategy.com) the 1234 sats from my LNBits instance with comment of this URL :-)
> 
> ```
> Created:
> 2023-04-21 00:27 (a few seconds ago)
> Expiry:
> 2023-04-22 00:27 (in a day)
> Description:
> Deposit to saylor@microstrategy.com
> Amount:
> -1234.000 sats
> Fee:
> 2.363 sats
> Payment hash:
> 0957fd449c7ef61ad5ae56db73813475844c1417f1dd8bfd024265fa5257191a
> Payment proof:
> 16bb95ee19206a0e688b218e1118f5ed9c2a6b52cf65510b247b4cecac9783fc
> comment:
> https://github.com/lnbits/lnbits/issues/1638
> ```

Yeah, great job. You figured it out yourself where to send if you want to send to Saylor? From the error message "undefined
503 SERVICE UNAVAILABLE"? Certainly not, I was telling you where to look, where is the problem. Without me, you, -as a great developer- would have spent probably a lot more time, looking in the codes and realizing that a 301 (or any other type of) redirect is causing the issue.

Do you think my grandma or any other average user -if requested to send some sats to saylor@microstrategy.com- will just simply realize from that error message that he / she needs to send in fact to saylor@www.microstrategy.com?

> Of course, it would be better if LUD-16 described more clearly whether redirects can be used or not. In clearer standards, such as RFC, such details are important. Everything is written here in general, so everyone interprets and implements as he wants. For example, you offered a maximum of 3 redirects, but why so many? Why not 5, or 10? And what if others implement a maximum of 5, others 3, and someone 1? In all cases, there will be situations where everything will work differently. In this context, LNBits has the number 0. What number other wallets that allow redirects have is known only to them.

Exactly: LUD-16 doesn't say anywhere that it is forbidden / not allowed to follow redirects. In this particular case, I think common sense in UX is to follow whatever is instructed by the original .well-known URL. In fact it should be like with email. I control myemail@mysite1.com, who can stop me to forward all these e-mails automatically to myemail@mysite2.com?

**Comment by dni** [comment]
is this resolved :)?

**Comment by Perlover** [comment]
As for me, I only answered here and did not create this ticket. But I just tested version `0.10.8` (commit f5343791e4366b7c3461073caab2bd133982a98c), which I have. I tested with the address that the ticket creator gave as an example. Address `saylor@microstrategy.com`

I'm getting error `503` right now.

But I believe, as I have already described above, this is not a mistake/bug. I'm not a developer. I want to state just a fact - the URI `https://microstrategy.com/.well-known/saylor` redirects through 301 HTTP Response to `https://www.microstrategy.com/.well-known/saylor`. The [LUD-16](https://github.com/lnurl/luds/blob/luds/16.md) does not mention anything about redirects and how to behave in such cases. Therefore, purely formally, the behavior of LNBits is not a bug. But let it be decided by the developers.
